publisher/publisher.go

```diff
@@ -117,20 +169,32 @@ func (pub *Impl) SubmitToSingleCT(ctx context.Context, logID string, der []byte)
 
 	chain := append([]ct.ASN1Cert{der}, pub.issuerBundle...)
 
-	ctLog := pub.logIDToLog(logID)
-	if ctLog == nil {
-		err := fmt.Errorf("Failed to find log matching ID: %s", logID)
-		pub.log.AuditErr(err.Error())
-		return err
+	// Try to find a *Log from the cache for this logURL. If this is the first
+	// request to submit to this log then add it to the cache
+	var ctLog *Log
+	if ctLog = pub.ctLogsCache.GetLog(logPublicKey); ctLog == nil {
+		var err error
+		ctLog, err = pub.ctLogsCache.AddLog(logURL, logPublicKey)
+		if err != nil {
+			pub.log.AuditErr(
+				fmt.Sprintf("Failed to add log to cache, logURL or logPublicKey malformed: %s",
```

Comment:
I would make this error "Making Log: %s", since the two possible sub-errors will show up in the %s.